### PR TITLE
Implementation and elaboration for MACRO_ARITH_STRING_PRED_ENTAIL

### PR DESCRIPTION
**Pull request by ajreynol**
This is the most widely applicable proof reconstruction tactic for strings.

Includes changes from https://github.com/cvc5/cvc5/pull/11073, review that first.

### CODE REVIEW

File: src/theory/arith/arith_rewriter.cpp [hunk 1]
```diff
@@ -75,6 +78,41 @@ Node ArithRewriter::rewriteViaRule(ProofRewriteRule id, const Node& n)
...
+      }
+      // first do basic length intro, which rewrites (str.len (str.++ x y))
+      // to (+ (str.len x) (str.len y))
+      Node nexp = ae.rewriteLengthIntro(tgt);
+      // Also must make this is a "simple" check (isSimple=true).
+      Node ret = ae.rewritePredViaEntailment(nexp, true);
+      if (!ret.isNull())
+      {
+        return ret;
+      }
```
Comment (abdoo8080) [inline src/theory/arith/arith_rewriter.cpp:113]:
```suggestion
      return ret;
```

File: src/theory/arith/arith_rewriter.cpp [hunk 2]
```diff
@@ -75,6 +78,41 @@ Node ArithRewriter::rewriteViaRule(ProofRewriteRule id, const Node& n)
...
+        tgt = ae.normalizeGeq(n);
+      }
+      if (tgt.isNull() || !tgt[0].getType().isInteger())
+      {
+        return Node::null();
+      }
+      // first do basic length intro, which rewrites (str.len (str.++ x y))
+      // to (+ (str.len x) (str.len y))
+      Node nexp = ae.rewriteLengthIntro(tgt);
+      // Also must make this is a "simple" check (isSimple=true).
```
Comment (abdoo8080) [inline src/theory/arith/arith_rewriter.cpp:108]:
```suggestion
      // Also must make this a "simple" check (isSimple=true).
```

File: src/theory/arith/arith_rewriter.cpp [hunk 3]
```diff
@@ -75,6 +78,41 @@ Node ArithRewriter::rewriteViaRule(ProofRewriteRule id, const Node& n)
       }
     }
     break;
+    case ProofRewriteRule::MACRO_ARITH_STRING_PRED_ENTAIL:
+    {
+      // only matters if contains strings
```
Comment (abdoo8080) [inline src/theory/arith/arith_rewriter.cpp:83]:
```suggestion
      // only matters if n contains strings
```

**Review by abdoo8080 — APPROVED** [review]
LGTM! Left some comments though.

File: src/rewriter/basic_rewrite_rcons.cpp [hunk 4]
```diff
@@ -219,6 +225,221 @@ bool BasicRewriteRCons::ensureProofMacroBoolNnfNorm(CDProof* cdp,
...
+  }
+  if (approxRew != approx)
+  {
+    Node areq = approxGeq.eqNode(approxRewGeq);
+    Trace("brc-macro") << "- prove " << areq << " via arith-poly-norm"
+                       << std::endl;
+    if (!ensureProofArithPolyNormRel(cdp, areq))
+    {
+      Trace("brc-macro") << "...failed to show normalization" << std::endl;
+      return false;
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/rewriter/basic_rewrite_rcons.cpp:334]: Actually I will do `Assert(false)` (an accidental proof hole should only be a failure in debug).

File: src/rewriter/basic_rewrite_rcons.cpp [hunk 5]
```diff
@@ -219,6 +225,221 @@ bool BasicRewriteRCons::ensureProofMacroBoolNnfNorm(CDProof* cdp,
...
+        {
+          Node eqc = i == 0 ? eqii[0][i].eqNode(aec) : aec.eqNode(eqii[0][i]);
+          cdp->addStep(eqc, ProofRule::ARITH_POLY_NORM, {}, {eqc});
+          transEq.push_back(eqc);
+        }
+      }
+      if (transEq.size() == 2)
+      {
+        cdp->addStep(eqii[0], ProofRule::TRANS, transEq, {});
+      }
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/rewriter/basic_rewrite_rcons.cpp:293]: Yeah I think you're right...

**Pull request merged**